translator/src/translator.rs

```diff
@@ -0,0 +1,450 @@
+            };
+            let callable_type = types::CallableType::new(
+                old_callable.name,
+                old_callable.signature[1..].to_vec(),
+                old_callable.return_type,
+                old_callable.is_async,
+            );
+            typ = PythonType::Callable(Box::new(callable_type));
+        }
+        match typ {
+            PythonType::Callable(callable) => {
+                let mut num_pos_args = 0;
+                // First check all the positional args
+                for (i, arg) in callable.signature.iter().enumerate() {
+                    match arg {
+                        types::CallableArgs::Positional(t) => {
+                            self.check_type(&c.args[i].get_node(), t);
+                            if i != 0 {
+                                write!(self.output, ", ");
+                            }
+                            self.visit_expr(&c.args[i]);
+                            num_pos_args = i;
+                        },
+                        types::CallableArgs::Args(_t) => {
+                            break;
+                        },
+                        _ => {}
+                    }
+                }
+                // Then check all the star args if there are any
+                if num_pos_args + 1 < c.args.len() {
+                    write!(self.output, "{{");
+                    for (i, arg) in c.args[num_pos_args..].iter().enumerate() {
+                        self.check_type(&arg.get_node(), callable.signature[num_pos_args+i].get_type());
+                        if i != 0 {
+                            write!(self.output, ", ");
+                        }
+                        self.visit_expr(arg);
+                    }
+                    write!(self.output, "}}");
+                }
+            },
+            _ => {
+                println!("Shouldn't hit this code path");
+            }
+        }
+        // for keyword in &c.keywords {
+        //     self.visit_expr(&keyword.value);
+        // }
+        write!(self.output, ")");
+    }
+
+    fn visit_attribute(&mut self, attribute: &Attribute) {
+        self.visit_expr(&attribute.value);
+        match &attribute.value {
+            Expression::Name(n) =>  {
+                let symbol_table = self.checker.get_symbol_table(None);
+                match symbol_table.lookup_in_scope(&n.id, self.current_scope) {
+                    Some(entry) => {
+                        match entry.last_declaration() {
+                            symbol_table::Declaration::Alias(_a) => {
+                                write!(self.output, "::{}", attribute.attr);
+                                return
+                            },
+                            _ => {}
+                        }
+                    },
+                    None => {},
+                }
+            },
+            _ => {}
+        }
+        write!(self.output, ".{}", attribute.attr);
```

Comment:
does this also need to handle outputting "->" too? when its a pointer?

Reply:
This is an open question, right now I was trying to avoid pointers (see e.g. the "auto self& = *this;" hack) -- in most cases it will be great to have `value types` like std::string or std::vector or torch::Tensor that are small proxy objects and manage their memory internally (maybe using smart pointers internally). But the details will need to be figured out.